### ISSUE TITLE
# [Bug] `tile.cast` and `tile.row_sum`  get wrong  `valid_shape`

### ISSUE THREAD

**Issue by xzhxzhxzh123** [hw-native-sys/pypto#1401]
### Component

Parser

## 运行环境

| 项 | 值 |
|----|-----|
| pypto commit | `7c207b65` — fix(simplify): fold dead if-guards keyed on loop-affine scalars (#1391) |
| pypto-lib commit | `923ef36` — Fix CSA indexer intermediate reuse (#305) |
| ptoas-bin | v0.40 |
| Python | 3.10.20 |
| 设备 | Ascend a2a3

## 启动命令

```bash
  python pypto-lib/models/qwen3/14b/qwen3_14b_decode_4core_nofillpad.py \
  -d 11 --enable-l2-swimlane
```

## 输出路径

```
build_output/Qwen3Decode_20260519_091555/
├── ptas/rmsnorm_kernel.pto    ← valid_row 对比的关键文件
├── report/
└── ...
```

### 现象

Qwen3-14B decode RMSNorm 切 4 SPMD 核时，全部 81920 个输出元素为 NaN：

```
  'out' FAIL  shape=(16, 5120) dtype=torch.bfloat16
    Mismatched elements: 81920/81920
    [0] actual=nan, expected=-0.08251953125
    ...
```

### DSL 层面

文件：`qwen3_14b_decode_4core_nofillpad.py`

每核 load `[8, 256]` 的 tile，但仅前 `[local_valid, 256]` 行有效（`local_valid ≤ 4`）。期望硬件仅对有效行执行 `tmul` / `trowsum`：

```python
x_chunk = pl.cast(
    pl.slice(hidden_states, [8, 256], [b0+row_start, k0],
             valid_shape=[local_valid, 256]),
    target_type=pl.FP32,
)
partial_sq = pl.add(partial_sq,
    pl.reshape(pl.row_sum(pl.mul(x_chunk, x_chunk)), [1, 8]))
```

### ptoas 层面 — valid_row 传播断裂

```
pto 指令       | alloc_tile 的 valid_row     | 来源 / 问题
---------------|----------------------------|-------------
pto.tload      | valid_row = %55  (dynamic) | ✓ 上游 load 正确携带 valid_shape = [local_valid, 256]
pto.tcvt       | valid_row = %c8_index      | ✗ 丢失！写死为物理行数 8，不再引用动态 %55
pto.tmul       | valid_row = %c8_index      | ✗ 从 tcvt 继承物理值 8
pto.trowsum    | valid_row = %c8_index      | ✗ 从 tmul 继承物理值 8
```

**对照：** `pto.tload` → `valid_row = %55`（`%55` 是 `local_valid` 的动态 SSA 值）。**经过 `pto.tcvt` 后变成常量 `%c8_index`（物理行数 8）**，不再表示 `[local_valid, 1]`。

#### Bug 1（根因）— `tile.cast` type deduction 不传播 valid_shape

**文件：** `pypto/src/ir/op/tile_ops/unary.cpp`

**行号：144**

```cpp
  // Cast operation preserves shape but changes data type
  TileView tile_view;
  tile_view.valid_shape = tile_type->shape_;  // ← BUG: 写死为物理 shape
  InheritTileViewLayout(tile_view, tile_type);
  return std::make_shared<TileType>(tile_type->shape_, target_dtype, std::nullopt, tile_view);
```

`tile_type->shape_` 返回物理 shape `[8, 256]`。**输入 tile 的 `tile_view_->valid_shape`（即 `[local_valid, 256]`）被完全忽略。**

#### Bug 2 — `tile.row_sum` type deduction 不传播 valid_shape

**文件：** `pypto/src/ir/op/tile_ops/reduction.cpp`

**行号：154**

```cpp
  // Output shape is [...batch_dims, rows, 1]
  std::vector<ExprPtr> output_shape(input_shape.begin(), input_shape.end() - 1);
  output_shape.push_back(ConstInt(1, INDEX));
  TileView tile_view;
  tile_view.blayout = TileLayout::col_major;
  tile_view.valid_shape = output_shape;  // ← BUG: 写死为物理输出 shape
```

即使 cast 修复后 row_sum 能收到正确的输入 valid_shape，这里也会再次丢失——`output_shape` 永远是物理 shape `[8, 1]`，不包含 `[local_valid, 1]`。

### 正确参考实现：`tile.mul` type deduction

**文件：** `pypto/src/ir/op/tile_ops/elementwise.cpp`

**行号：130-132**

```cpp
  TileView tile_view;
  tile_view.valid_shape = GetValidShape(tile_type1);  // ← 正确：从输入继承 valid_shape
  InheritTileViewLayout(tile_view, tile_type1);
```

核心 helper：

**文件：** `pypto/src/ir/op/tile_ops/elementwise.cpp`

**行号：46-51**

```cpp
static std::vector<ExprPtr> GetValidShape(const shared_ptr<const TileType>& tile_type) {
  if (tile_type->tile_view_ && !tile_type->tile_view_->valid_shape.empty())
    return tile_type->tile_view_->valid_shape;  // 优先返回 valid_shape
  return tile_type->shape_;                     // fallback 物理 shape
}
```

## 根因分析

1. **`tile.cast`（根因）** — `DeduceTileCastType` 在 `unary.cpp:144` 写死 `valid_shape = tile_type->shape_`（物理 shape），完全未访问输入的 `tile_view_->valid_shape`。

2. **`tile.row_sum`（次因）** — `DeduceTileRowReductionType` 在 `reduction.cpp:154` 同样写死 `valid_shape = output_shape`。即使 bug 1 修复后输入 valid_shape 可达，这里也会二次丢失。

3. 对比 `tile.mul` — 它的 type deduction 通过 `GetValidShape(tile_type1)` 正确继承了 valid_shape。`GetValidShape` helper 已存在，框架的设计意图就是优先传播 valid_shape。

4. **后端无问题** — `ComputeAllocTileFields`（`pto_codegen.cpp:854-866`）正确读取 `tile_view_->valid_shape` 并生成 ptoas 的 `valid_row` operand。硬件 `tmul` 已验证会根据 `valid_row` 限制计算行数。**断在 IR 层 type deduction 没写入。**

### 修复 1：`tile.cast` — `DeduceTileCastType`

`pypto/src/ir/op/tile_ops/unary.cpp:144`

```diff
-  tile_view.valid_shape = tile_type->shape_;
+  tile_view.valid_shape = GetValidShape(tile_type);
```

### 修复 2：`tile.row_sum` — `DeduceTileRowReductionType`

`pypto/src/ir/op/tile_ops/reduction.cpp:148-155`

```diff
+  const auto& iv = tile_type->tile_view_;
+  const auto& input_valid = (iv.has_value() && !iv->valid_shape.empty())
+      ? iv->valid_shape : input_shape;
   TileView tile_view;
   tile_view.blayout = TileLayout::col_major;
-  tile_view.valid_shape = output_shape;
+  if (!input_valid.empty()) {
+    std::vector<ExprPtr> output_valid(input_valid.begin(), input_valid.end() - 1);
+    output_valid.push_back(std::make_shared<ConstInt>(1, DataType::INDEX, Span()));
+    tile_view.valid_shape = output_valid;
+  } else {
+    tile_view.valid_shape = output_shape;
+  }
```

### 修复 3（同类）：`tile.col_sum` — `DeduceTileColReductionType`

`pypto/src/ir/op/tile_ops/reduction.cpp:176-182` 存在相同问题，同上修复。

## 复现文件索引

| 用途 | 路径 |
|------|------|
| 触发 NaN 版本（无 fillpad） | `pypto-lib/models/qwen3/14b/qwen3_14b_decode_4core_nofillpad.py` |
| 工作版本（有 fillpad，冗余计算） | `pypto-lib/models/qwen3/14b/qwen3_14b_decode.py` |
| 详细分析文档 | `pypto-lib/models/qwen3/14b/bug_report_valid_shape_propagation.md` |
| ptoas 证据 | `build_output/Qwen3Decode_20260519_091555/ptoas/rmsnorm_kernel.pto` |

---

**总结：框架已提供 `GetValidShape` helper 和 `ComputeAllocTileFields` 的正确链路，问题仅在于 `tile.cast`（unary.cpp:144）和 `tile.row_sum`（reduction.cpp:154）的 type deduction 未调用该 helper。属于同类模式下的遗漏 bug，不是合理设计。**

### Steps to Reproduce

```markdown
见描述
```

### Expected Behavior

脚本能够正常通过

### Actual Behavior

出现Nan精度问题

### Git Commit ID

7c207b65

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

**Comment by luohuan19** [comment]
## Additional evidence: 9c64ded already landed "fix 1"; "fix 2" gap is now triggering Qwen3-14B `--l3` regression

Just hit the same root cause on a different path. A few clarifications and supporting data:

### Current `unary.cpp:50` state ≠ what this issue shows

The "Bug 1 (根因)" snippet in the description is from **before** PR #1395. That PR was merged as `9c64ded9` on 2026-05-18:

```cpp
// pypto/src/ir/op/tile_ops/unary.cpp@HEAD (current)
TileView tile_view;
tile_view.valid_shape = GetValidShape(tile_type);    // ← already fixed
InheritTileViewLayout(tile_view, tile_type);
return std::make_shared<TileType>(tile_type->shape_, tile_type->dtype_, std::nullopt, tile_view);
```

So **"fix 1" proposed in this issue is already in `main`**. What remains is exactly the "fix 2" (row_sum / col_sum) — and the half-merged state is now actively breaking downstream code, because IR producers (cast) propagate valid_shape but IR consumers (row_sum, matmul) still ignore it. The LB region between them is now unwritten.

### Independent trigger path: Qwen3-14B L3 generate end-to-end

Independent of `qwen3_14b_decode_4core_nofillpad.py` from the description, the released `qwen3_14b_npu_generate.py --l3` path also hits this, with all generated tokens collapsing to `0`:

```
$ python llm/examples/qwen3_14b_npu_generate.py --model-dir <qwen3-14b> \
    --prompt 'Huawei is' --platform a2a3 --max-seq-len 512 \
    --max-new-tokens 2 --device-id 0 --l3
...
text: !!
token_ids: [0, 0]
```

Same chain that #1401 identifies:

```
pl.slice(valid_shape=[cur_valid=1, K])         →  validRow=1 hint
  ↓ pl.cast (POST-9c64ded: inherits)              writes LB row 0 only
  ↓ pl.mul   (binary, inherits LHS)               writes LB row 0 only
  ↓ pl.row_sum (STATIC output_shape, unchanged)   reads LB rows 0..15
                                                  rows 1..15 = stale LB residue
  ↓ partial_sq accumulates garbage on lanes 1..15
  ↓ ... propagates through 40 decode layers, magnified by RMSNorm + matmul
  ↓ logits → NaN
  ↓ argmax(NaN) = 0
```

Sites hit in pypto-lib (commit `a5ab419` Add kernel insight export for qwen3 decode):

| File | Sites |
|---|---|
| `models/qwen3/14b/qwen3_14b_l3_generate.py` | scope1 rmsnorm ×2, prefill out_proj_residual, decode rmsnorm ×2, decode out_proj_residual |
| `models/qwen3/14b/qwen3_14b_decode_full.py` | scope1 rmsnorm ×2, scope3 out_proj_residual |
| `models/qwen3/14b/qwen3_14b_prefill.py` | rmsnorm ×2, out_proj_residual |

All match the pattern `pl.cast(pl.slice(..., valid_shape=[cur_valid_or_valid_tok, ...]), pl.FP32)` consumed by `pl.row_sum` / `pl.matmul` downstream.

### Micro-kernel minimal repro

Smaller than the qwen3_14b 4-core path — isolates exactly the failing chain:

```python
"""IR-level repro: cast inherits valid_shape but row_sum forces static.

Expected (intent): TROWSUM reads validRow = valid_rows (dynamic)
Actual (post-9c64ded): TROWSUM reads validRow = 16 (static),
                       picks up LB rows that TMUL never wrote.
"""
import pypto.language as pl

H_STATIC = 16
K_STATIC = 32

@pl.program
class CastRowsumMismatch:
    @pl.function(type=pl.FunctionType.Opaque)
    def repro(
        self,
        x: pl.Tensor[[H_STATIC, K_STATIC], pl.BF16],
        valid_rows: pl.Tensor[[1], pl.INT32],
        out: pl.Out[pl.Tensor[[1, H_STATIC], pl.FP32]],
    ) -> pl.Tensor[[1, H_STATIC], pl.FP32]:
        vr = pl.cast(pl.tensor.read(valid_rows, [0]), pl.INDEX)
        sliced = pl.slice(
            x, [H_STATIC, K_STATIC], [0, 0],
            valid_shape=[vr, K_STATIC],
        )
        casted = pl.cast(sliced, target_type=pl.FP32)   # validRow = vr (post-9c64ded)
        sq = pl.mul(casted, casted)                     # validRow = vr (binary inherits)
        sums = pl.row_sum(sq)                           # validRow = H_STATIC (BUG — see reduction.cpp:154)
        flat = pl.reshape(sums, [1, H_STATIC])
        out = pl.assemble(out, flat, [0, 0])
        return out
```

Dump the lowered pto IR to see the mismatch directly:

```
pto.tcvt    valid_row = %vr           # dynamic, correct
pto.tmul    valid_row = %vr           # dynamic, correct
pto.trowsum valid_row = %c16_index    # static, WRONG (loses %vr)
```

With `valid_rows=[1]` and `x[0,:]=1.0, x[1..15,:]=0.0`:

- **Expected**: `out[0, 0] = K_STATIC = 32`, `out[0, 1..15] = 0` (since x is 0 there, OR pre-fix unconditional full TCVT writes 0 there too)
- **Actual**: `out[0, 0] = 32`, `out[0, 1..15] = sum(LB_residue²)` — varies by LB allocator state; typically non-zero and may overflow to inf/NaN under load.

### Lib-side workaround currently in use

While "fix 2" is pending, the lib wraps `pl.fillpad(..., pad_value=pl.PadValue.zero)` around every `pl.cast(pl.slice(..., valid_shape=...))` site to explicitly promote valid_shape back to static and zero the padding. This restores pre-9c64ded behavior (cast writes the full tile) without depending on GM happening to be zero.

```python
# Before (broken post-9c64ded):
x_chunk = pl.cast(
    pl.slice(current_hidden, [BATCH_TILE, K], [b0, k0],
             valid_shape=[cur_valid, K]),
    target_type=pl.FP32,
)

# After (workaround):
x_chunk = pl.fillpad(
    pl.cast(
        pl.slice(current_hidden, [BATCH_TILE, K], [b0, k0],
                 valid_shape=[cur_valid, K]),
        target_type=pl.FP32,
    ),
    pad_value=pl.PadValue.zero,
)
```

This is purely defensive — it costs one extra `TFILLPAD` per chunk (negligible) but should not be needed once `row_sum` / `col_sum` / `matmul` consume valid_shape correctly per this issue's "fix 2".

### Ask

Please prioritize "fix 2" (`DeduceTileRowReductionType` propagating input valid_shape) plus the analogous `tile.col_sum` path mentioned in the description. Also worth auditing whether `pl.matmul` should propagate LHS valid_shape on its M dim — the `out_proj_residual` site relies on binary `pl.add` of `matmul_out` (static) + `cast_resid` (cur_valid), which is currently a "static wins" path that masks part of the problem but not the row_sum path.

Happy to test a candidate patch on the Qwen3-14B `--l3` end-to-end and report whether tokens come back.
